compiler/formatter/src/format.rs

```diff
@@ -0,0 +1,1704 @@
+    use candy_frontend::{rcst_to_cst::RcstsToCstsExt, string_to_rcst::parse_rcst};
+    use itertools::Itertools;
+
+    // Comments with code snippets display the formatted/expected version of the subsequent test,
+    // excluding a trailing newline.
+    //
+    // They're present for multiline expressions for better readability. When multiple source
+    // expressions have the same formatted version, there's only a comment in front of the first
+    // test case.
+
+    #[test]
+    fn test_csts() {
+        test(" ", "");
+        test("foo", "foo\n");
+        test("foo\n", "foo\n");
+
+        // Consecutive newlines
+
+        // foo
+        // bar
+        test("foo\nbar", "foo\nbar\n");
+        // foo
+        //
+        // bar
+        test("foo\n\nbar", "foo\n\nbar\n");
+        // foo
+        //
+        //
+        // bar
+        test("foo\n\n\nbar", "foo\n\n\nbar\n");
+        test("foo\n\n\n\nbar", "foo\n\n\nbar\n");
+        test("foo\n\n\n\n\nbar", "foo\n\n\nbar\n");
+        // foo = bar
+        //
+        // baz
+        test("foo =\n  bar\n\nbaz", "foo = bar\n\nbaz\n");
+        // foo = bar
+        //
+        // # abc
+        test("foo =\n  bar\n\n# abc", "foo = bar\n\n# abc\n");
+
+        // Consecutive expressions
+
+        // foo
+        // bar
+        // baz
+        test("foo\nbar\nbaz", "foo\nbar\nbaz\n");
+        // foo
+        // bar
+        test("foo\n bar", "foo\nbar\n");
+        // foo
+        //
+        // bar
+        test("foo\n \nbar", "foo\n\nbar\n");
+        // foo
+        test("foo ", "foo\n");
+
+        // Leading newlines
+
+        test(" \nfoo", "foo\n");
+        test("  \nfoo", "foo\n");
+        test(" \n  \n foo", "foo\n");
+
+        // Trailing newlines
+
+        test("foo\n ", "foo\n");
+        test("foo\n  ", "foo\n");
+        test("foo \n  ", "foo\n");
+        test("foo\n\n", "foo\n");
+        test("foo\n \n ", "foo\n");
+
+        // Comments
+
+        // # abc
+        // foo
+        test("# abc\nfoo", "# abc\nfoo\n");
+        // foo # abc
+        test("foo# abc", "foo # abc\n");
+        test("foo # abc", "foo # abc\n");
+        test("foo # abc ", "foo # abc\n");
+        // foo
+        // # abc
+        test("foo\n# abc", "foo\n# abc\n");
+        test("foo\n # abc", "foo\n# abc\n");
+    }
+    #[test]
+    fn test_int() {
+        test("1", "1\n");
+        test("123", "123\n");
+    }
+    #[test]
+    fn test_binary_bar() {
+        test("foo | bar", "foo | bar\n");
+        test("foo|bar", "foo | bar\n");
+        test("foo  |  bar", "foo | bar\n");
+        test("foo\n\n|   bar", "foo | bar\n");
+        test("foo | (bar)", "foo | bar\n");
+        test("foo | (\n  bar\n)", "foo | bar\n");
+        test("foo | (bar baz)", "foo | (bar baz)\n");
+        test("foo | (bar | baz)", "foo | (bar | baz)\n");
+        test(
+            "veryVeryVeryVeryVeryVeryVeryVeryLongReceiver | (veryVeryVeryVeryVeryVeryVeryVeryVeryVeryLongFunction)",
```

Comment:
For easier readability, I'd rename these to `loooooooooooooooooooooooooongReceiver` and similar